### PR TITLE
allow as gen_server

### PR DESCRIPTION
**Pull request by sparrell**

### CODE REVIEW

File: apps/ocas/src/action_servers/act_allow.erl [hunk 1]
```diff
@@ -33,13 +33,81 @@
...
+
+%% interface calls
+-export([ start/1
+        , stop/0 
+        , keepalive/0 
+        ]).
 
 %% This routine API handles all the actions that can be taken
 
+start(State) ->
```
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_allow.erl]: According to **Elvis**:
> Line 60 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl [hunk 2]
```diff
@@ -33,13 +33,81 @@
...
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+%% interface calls
+-export([ start/1
+        , stop/0 
+        , keepalive/0
```
Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 55 has 1 trailing whitespace characters.

File: test/allow_SUITE.erl [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                    %% key exists and value correct so go to next
+                    check_map( RestOfExpectedJsonPairs, JsonMap);
+                BadValue ->
+                    %% key exists but has wrong value so format test fail error
+                    ErrorText = list_to_binary(io_lib:format("~p is bad value for key ~p", [BadValue, Key])),
+                    cause_failure = ErrorText
+             end;
+        false ->
+            %% key doesn't exist so format test fail error
+            ErrorText = list_to_binary(io_lib:format("key = ~p not found in json", [Key])),
```
Comment (gadgetci) [inline test/allow_SUITE.erl]: According to **Elvis**:
> Line 233 is too long:             ErrorText = list_to_binary(io_lib:format("key = ~p not found in json", [Key])),.

Comment (elvisci) [inline test/allow_SUITE.erl]: Line 233 is too long:             ErrorText = list_to_binary(io_lib:format("key = ~p not found in json", [Key])),.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 2 above)
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_allow.erl]: According to **Elvis**:
> Line 55 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl [hunk 4]
```diff
@@ -33,13 +33,81 @@
...
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+%% interface calls
+-export([ start/1
+        , stop/0
```
Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 54 has 1 trailing whitespace characters.

File: apps/ocas/src/openc2_handler.erl [hunk 5]
```diff
@@ -153,6 +153,37 @@ check_valid_action( false, _ActionModule, _ActionFunction, Req, State ) ->
...
+    ModifiersKeyExists = maps:is_key( <<"modifiers">>, JsonMap ),
+    State4 = maps:put(has_modifiers, ModifiersKeyExists, State3),
+
+    %% spin up the process
+    ActionPid = ActionModule:start(State4),
+    lager:debug("got past ~p spin up ~p", [ActionModule, ActionPid]),
+    lager:debug("State4: ~p ", [State4]),
+
+    %% test keep alive - should get {keepalive_received, allow_server}
+    ActionKeepAlive = ActionModule:keepalive(),
```
Comment (gadgetci) [inline apps/ocas/src/openc2_handler.erl]: According to **Elvis**:
> Remove the dynamic function call on line 181. Only modules that define callbacks should make dynamic calls.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 1 above)
Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 60 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 4 above)
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_allow.erl]: According to **Elvis**:
> Line 54 has 1 trailing whitespace characters.

File: test/allow_SUITE.erl [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    case maps:is_key(Key, JsonMap) of
+        true ->
+            %% key exists so check value
+            case maps:get(Key, JsonMap) of
+                ExpectedValue ->
+                    %% key exists and value correct so go to next
+                    check_map( RestOfExpectedJsonPairs, JsonMap);
+                BadValue ->
+                    %% key exists but has wrong value so format test fail error
+                    ErrorText = list_to_binary(io_lib:format("~p is bad value for key ~p", [BadValue, Key])),
```
Comment (elvisci) [inline test/allow_SUITE.erl]: Line 228 is too long:                     ErrorText = list_to_binary(io_lib:format("~p is bad value for key ~p", [BadValue, Key])),.

File: apps/ocas/src/openc2_handler.erl [hunk 7]
```diff
@@ -153,6 +153,37 @@ check_valid_action( false, _ActionModule, _ActionFunction, Req, State ) ->
...
+    JsonMap = maps:get(json_map, State),
+    TargetKeyExists = maps:is_key( <<"target">>, JsonMap ),
+    State2 = maps:put(has_target, TargetKeyExists, State),
+    ActuatorKeyExists = maps:is_key( <<"actuator">>, JsonMap ),
+    State3 = maps:put(has_actuator, ActuatorKeyExists, State2),
+    ModifiersKeyExists = maps:is_key( <<"modifiers">>, JsonMap ),
+    State4 = maps:put(has_modifiers, ModifiersKeyExists, State3),
+
+    %% spin up the process
+    ActionPid = ActionModule:start(State4),
```
Comment (gadgetci) [inline apps/ocas/src/openc2_handler.erl]: According to **Elvis**:
> Remove the dynamic function call on line 176. Only modules that define callbacks should make dynamic calls.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 2 above)
Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 55 has 1 trailing whitespace characters.

File: test/allow_SUITE.erl (hunk 6 above)
Comment (gadgetci) [inline test/allow_SUITE.erl]: According to **Elvis**:
> Line 228 is too long:                     ErrorText = list_to_binary(io_lib:format("~p is bad value for key ~p", [BadValue, Key])),.

File: apps/ocas/src/openc2_handler.erl (hunk 5 above)
Comment (elvisci) [inline apps/ocas/src/openc2_handler.erl]: Remove the dynamic function call on line 181. Only modules that define callbacks should make dynamic calls.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 1 above)
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_allow.erl]: According to **Elvis**:
> Line 60 has 1 trailing whitespace characters.

Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 60 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_augment.erl [hunk 8]
```diff
@@ -33,33 +33,75 @@
...
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+%% interface calls
+-export([ start/1
+        , stop/0 
+        , keepalive/0
```
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_augment.erl]: According to **Elvis**:
> Line 53 has 1 trailing whitespace characters.

Comment (elvisci) [inline apps/ocas/src/action_servers/act_augment.erl]: Line 53 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl [hunk 9]
```diff
@@ -33,33 +33,75 @@
...
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+%% interface calls
+-export([ start/1
+        , stop/0 
+        , keepalive/0
```
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_allow.erl]: According to **Elvis**:
> Line 53 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_augment.erl [hunk 10]
```diff
@@ -33,33 +33,75 @@
...
+-export([ start/1
+        , stop/0 
+        , keepalive/0 
+        ]).
 
 %% This routine API handles all the actions that can be taken
 
-augment_server(State) ->
-    %% separate process to handle augment action
+start(State) ->
```
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_augment.erl]: According to **Elvis**:
> Line 58 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl [hunk 11]
```diff
@@ -33,33 +33,75 @@
...
+-export([ start/1
+        , stop/0 
+        , keepalive/0 
+        ]).
 
 %% This routine API handles all the actions that can be taken
 
-allow_server(State) ->
-    %% separate process to handle allow action
+start(State) ->
```
Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 58 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_augment.erl [hunk 12]
```diff
@@ -33,33 +33,75 @@
...
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+%% interface calls
+-export([ start/1
+        , stop/0
```
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_augment.erl]: According to **Elvis**:
> Line 52 has 1 trailing whitespace characters.

Comment (elvisci) [inline apps/ocas/src/action_servers/act_augment.erl]: Line 52 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 11 above)
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_allow.erl]: According to **Elvis**:
> Line 58 has 1 trailing whitespace characters.

Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 58 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_augment.erl (hunk 10 above)
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_augment.erl]: According to **Elvis**:
> Line 58 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_augment.erl (hunk 12 above)
Comment (elvisci) [inline apps/ocas/src/action_servers/act_augment.erl]: Line 52 has 1 trailing whitespace characters.

Comment (gadgetci) [inline apps/ocas/src/action_servers/act_augment.erl]: According to **Elvis**:
> Line 52 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_augment.erl (hunk 10 above)
Comment (elvisci) [inline apps/ocas/src/action_servers/act_augment.erl]: Line 58 has 1 trailing whitespace characters.

Comment (elvisci) [inline apps/ocas/src/action_servers/act_augment.erl]: Line 58 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 9 above)
Comment (gadgetci) [inline apps/ocas/src/action_servers/act_allow.erl]: According to **Elvis**:
> Line 53 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl [hunk 13]
```diff
@@ -33,33 +33,75 @@
...
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+%% interface calls
+-export([ start/1
+        , stop/0
```
Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 52 has 1 trailing whitespace characters.

File: apps/ocas/src/action_servers/act_allow.erl (hunk 9 above)
Comment (elvisci) [inline apps/ocas/src/action_servers/act_allow.erl]: Line 53 has 1 trailing whitespace characters.

**Pull request merged**